[RFC] Only allow encoding=utf-8 and simplify multibyte code

bfredl opened the pull request:
During #2469 the possibilty to only allow `encoding=utf-8` was discussed, which would enable simplifications both for nvim core, and for guis/plugins, which then always can assume the encoding to be utf-8.

`&encoding` has several effects, like being the default for `setglobal fileencoding` (encoding of new files), and in nvim the TUI (currently) always uses it. The internal effect is that switches between three different modes in which text processing is done: 1-byte "latin", 2-byte "dbcs", and unicode "utf8", by setting flags and reassigning a number of function pointers. A quick grep shows that these flags (`enc_utf8|has_mbyte|enc_dbcs|enc_latin1like`) are used in 600 places throughout the codebase. (Though it would be unrealistic to remove them all in one PR, and if done it should be done with some care, as they do serve as documentation for what assumtions the code is relying on)

AFAIK nowadays almost everyone is using `encoding=utf-8` and I believe that any text processing in latin and dbcs mode could just as well be done by converting to utf-8 and using utf-8 mode, and whatever performance difference  should be pretty small in the large context. Please correct me if this isn't true.

What this PR does so far is always set `p_enc` to "utf-8" at start up. `set encoding=utf-8` (and synonyms) are allowed (as it is very common in vimrc:s), but any other value is an error. I've started cleaning up `mbyte.c`.

What might should be done is use `enc_locale()` as a default for `setglobal fenc`, to preserve the effect locale has on encoding for new files. Or perhaps just leave the default with `utf-8`, not sure what's best.
Perhaps also should implement conversion from/to locale encoding (if not utf-8) in the TUI. This might not be widely used, but it should not be too much work to support at least 1-byte encodings. (no buffering of leftover bytes and such).

There is more cleanup to be done (not to mention documentation) though it would be best to discuss if we really want to do this before doing more mork.

bfredl committed:
encoding: only allow encoding=utf-8

bfredl committed:
encoding: cleanup mbyte.c given fixed encoding=utf-8

Eliminate mb_init():
Set "enc_utf" and "has_mbyte" early. Eliminate "enc_unicode" and "enc_latin1like".
init_chartab() and screenalloc() are already invoked elsewhere
in the initialization process.
The EncodingChanged autocmd cannot be triggered.
At initialization, there is no spellfiles to reload

bfredl committed:
encoding: delete non-UTF-8 implementations of multibyte functions

Deleted documentation was duplicated at specific utf_ implementation

justinmk commented:
Nice code reduction. This could be worth considering in the future. But in the short term[1], we should simply disallow `set encoding` *after* `VimEnter`. That is the simplest possible solution, it solves the problem reliably and with very little risk. 

I recently found a [2009 vim_dev thread](https://groups.google.com/d/msg/vim_dev/up1KslPRGzk/9xwHeIb6WWgJ) which makes it clear that there is zero reason for `'encoding'` to be user-configurable. (I believe @jamessan is the author of that thread, maybe he can offer his thoughts here in 2015.) The [entreaty](https://groups.google.com/d/msg/vim_dev/up1KslPRGzk/vfn0QWLKtacJ) for the current behavior instead makes the case *stronger* by raising objections that have no relevance to Neovim.

- concern about `--disable-multibyte`, a compile-time option which is absent in Neovim
- concern about `'termencoding'`, which was removed in Neovim
- concern about Windows 98 and 16-bit systems
- hypothetical "desire to create or edit files using characters not supported by your system locale, for instance multi-charset files in UTF-8 when the Windows locale is Windows-1252,"

And just to mention:

> changing the 'encoding' _after_ the end of startup, when you already have one or more buffers loaded, is not something I would recommend; it may lead to dataloss or file data corruption


 
---

[1] I.e., the next 2 months, where progress towards a `v0.1` release is of paramount importance--I am working on a roadmap which will contain more details.

bfredl commented:
Nice read. It also mentions using `++latin1` to emulate "single-byte" mode even if `encoding=utf-8`. I suppose the version after `0.1` would be the vNext milestone? I'll just leave this and open a separate PR for the `vimEnter` restriction.

bfredl commented:
Note to self: check/fix that`set fenc[s]=default` still works.

justinmk commented:
0.2. might a little too soon for this, as there are higher priorities and stability will be a chief concern there.

bfredl commented:
A possibility is to only merge the first commit at first (well, and the 2nd, but it's already with #2929 ), which has the full "operational" effect. But I don't really see the threat against stability. The following commits removing code in `mbyte.c` that then isn't needed. As I said above, the goal was NOT to change all the 600 places scattered around the codebase where there is a mbyte/non-mbyte branch in a single PR or fixed timeframe. (simpfications could be done "with some care" when a subsystem is being reworked anyway).

justinmk commented:
Ok. It's just that we are currently drowning in half-finished refactorings:

- remove char_u and friends
- include-what-you-use
- enable `-Wconversion`
- migrate legacy tests

And it's a cognitive burden trying to juggle all of these unfinished tasks, and then having to explain to newcomers the current state of things.

bfredl commented:
Perhaps. But isn't "explicitly" dead code (I mean the remaining non-mb code paths) better than implicitly almost-dead code that isn't tested neither by ci or by any developer( as our functional test suite won't even run with `encoding=latin1`.  `LC_ALL=en_US make test` just hangs) and very seldom is run, and thus easily could get undetected bugs. Now this code still has to be preserved and updated when working with an affected module, even if it isn't much tested, but with this PR it is allowed to be removed when it makes the work simpler.

justinmk commented:
Good point. Only thing blocking this for 0.2, then, is any opinion on whether there a reason to continue to allow changing `'encoding'`. @ZyX-I  @jamessan @chrisbra @Shougo @kopischke do you know of any major reason why users should need to change `'encoding'`?

ZyX-I commented:
@justinmk Changing `&encoding` at least at startup is definitely needed for testing ShaDa &encoding-to-utf-8 and back conversion.

Other than that I think it is only needed if

1. User for some reason needs batch processing of non-UTF-8 files, do this in NeoVim and not with something like perl and do it as fast as possible. (`set encoding={file encoding}` is faster then using `e ++enc={file encoding}` and then writing: saves two recodings; only noticeable on rather large files.) I am talking about *batch* processing because with regular editing you will hardly notice the difference.
2. There are ASCII-compatible encodings that have characters that are not present in Unicode. I do not know any of them though and am not sure that such a thing exists at all.

kopischke commented:
I agree the _technical_ reasons for having configurable `encoding` as described in [the Vim dev thread](https://groups.google.com/d/msg/vim_dev/up1KslPRGzk/9xwHeIb6WWgJ) are moot, obviates both by OS evolution and Neovim’s development. However, historically there have been grounds for text processing tools to support the internal representation of some Asian encodings even after Unicode was introduced, as the latter did not encode all of the former’s characters in its early incarnations.

I have no idea how much of that _internationalization_ reason applies to us, though, both in terms of software evolution (Unicode has steadily extended its character coverage since the standard’s inception) and of the technical abilities of Neovim’s Asian encoding support (does it actually handle characters excluded form UTF-8 representation?). It might be worth collecting some more opinions from Asian language users of Vim, not just Japanese, but also Chinese.

bfredl commented:
> Changing &encoding at least at startup is definitely needed for testing ShaDa &encoding-to-utf-8 and back conversion.

@ZyX-I But isn't this functionality needed precisely when `encoding!=utf-8` anyway ?

@kopischke Good point. I know that there have been controversy about the way UCS represents CJK characters (Han unification), but I don't know how much this applies to the latest versions, and if 2byte encodings are still used in vim for this reason.

justinmk commented:
@bfredl With this change, https://github.com/neovim/neovim/wiki/Troubleshooting#neovim-cant-read-utf-8-characters will be eliminated?

bfredl commented:
I didn't see a single issue there but more a number of environment encoding related problems. This won't magically remove all env encoding issues, but it would make the situation simpler (less "moving parts") and simpler to communicate: neovim *is* vim-with-encoding=utf8, always. It should at least eliminate the need for the workarounds `scriptencoding utf-8` in vimrc and override `LC_ALL` in a buggy locale environment (I mean when the locale "really" is utf-8 but the env vars fail to specify this, though we already have quite sane fallbacks with #935/#2469 )

justinmk commented:
@bfredl I am speaking specifically of this:

> If there's no results, then you might not be using a UTF-8 locale

My thought was: user's locale should not matter since we would no longer initialize `'encoding'` based on the locale. 

>  It should at least eliminate the need for the workarounds scriptencoding utf-8 in vimrc and override LC_ALL in a buggy locale environment

Those are good things to note.

bfredl commented:
> My thought was: user's locale should not matter since we would no longer initialize 'encoding' based on the locale. 

Correct. Though the regex on the wiki is incomplete (no results for me), I think it should be `locale | grep -E '(LANG|LC_CTYPE|LC_ALL)=(.*\.)?(UTF|utf)-?8'`

bfredl committed:
encoding: update tests

Change shada reencoding tests to check for
correct handling of UTF-8 and binary strings.

Delete enc=latin1 json tests.

bfredl committed:
encoding: update documentation

bfredl commented:
Marking RFC. There is certainly more cleanups to do, like in `charset.c` and `fileio.c`, but i think this PR does a reasonable amount of code change of removing the `mb_` function pointer and `MB_` macro indirections and simplifying the initialization, to be reviewed as a single PR. There might be some more doc work though, like `mbyte.txt` which contains lots of outdated info (also not related to this PR)

jamessan commented on src/nvim/screen.c:
```diff
@@ -5278,7 +5278,7 @@ void screen_puts_len(char_u *text, int textlen, int row, int col, int attr)
   int need_redraw;
 
-  const int l_has_mbyte = has_mbyte;
+  const bool l_has_mbyte = has_mbyte;
```
Isn't `has_mbyte` always true, thus allowing any code that relies on `l_has_mbyte` being false to be removed?

bfredl replied:
There are are lot's of references to `has_mbyte`, 388 in total on this branch, and it would not be reasonable to remove them all in one PR. `screen.c` is surely a module that could be simplified under the `enc=utf-8` assumption, but I rather do it in a separate PR where it will get sufficient attention by its own, after this one is merged. (though the addition of a lot more screen tests should make us a bit more confident of this sort of changes than when this PR was first proposed)
This PR needs review for the existing changes, rather than weighing it down with further changes spread across the codebase.

justinmk replied:
We could move forward with this after 0.1.5 is released?

bfredl replied:
Sure. This is anecdotal evidence of course, but given the complete lack of complaints like "why doesn't nvim work in my latin1 terminal anymore??" (well it still does, but you need to explicitly configure it now) or "why doesn't [some python or remote plugin] work with `enc=latin1`?" I think we should move forward.

nicolaiskogheim commented on test/functional/shada/variables_spec.lua:
```diff
@@ -91,35 +91,15 @@ describe('ShaDa support code', function()
-    set_additional_cmd('set encoding=latin1')
+    --set_additional_cmd('set encoding=latin1')
     reset()
     -- \171 is U+00AB LEFT-POINTING DOUBLE ANGLE QUOTATION MARK in latin1
```
This comment should be removed.

nicolaiskogheim commented on test/functional/shada/registers_spec.lua:
```diff
@@ -128,36 +128,25 @@ describe('ShaDa support code', function()
   function()
-    set_additional_cmd('set encoding=latin1')
     reset()
     -- \171 is U+00AB LEFT-POINTING DOUBLE ANGLE QUOTATION MARK in latin1
```
Same as with [this](https://github.com/neovim/neovim/pull/2905/files#r69023041). I believe the comment should be removed.

bfredl commented:
@nicolaiskogheim thanks, updated.

Rebased.

bfredl commented:
Maybe we should get this merged soon. There are already PRs that do/could make use of this like #5177 and #2598. Most changes are removals of conditions, indirections, and then unused functions, the big "non-trivial" change is the elimination of `mb_init()`. 

As said there is more clean up that can be done. I will do more in a follow-up PR, as it stands this PR is the starting point necessary for various PRs to clean-up different parts of the code base.

oni-link commented on src/nvim/mbyte.c:
```diff
@@ -1,68 +1,27 @@
+/// be one character.
+///
+/// utf-8 is used everywhere in the core. This is in registers, text
+/// manipulation, buffers, etc. Nvim core comminucates with external plugins
```
comminucates -> communicates

oni-link commented on src/nvim/mbyte.c:
```diff
@@ -600,10 +358,7 @@ int bomb_size(void)
   if (curbuf->b_p_bomb && !curbuf->b_p_bin) {
     if (*curbuf->b_p_fenc == NUL) {
       if (enc_utf8) {
```
If `enc_utf8` is always true, this condition can be removed.

oni-link commented on src/nvim/mbyte.c:
```diff
@@ -2239,13 +1751,7 @@ int mb_tail_off(char_u *base, char_u *p)
-
-  /* Return 1 when on the lead byte, 0 when on the tail byte. */
-  return 1 - dbcs_head_off(base, p);
+  return 0;
```
With `enc_utf8` always true this will never be reached.

oni-link commented on src/nvim/mbyte.c:
```diff
@@ -2461,12 +1967,9 @@ int mb_fix_col(int col, int row)
-          && dbcs_screen_head_off(ScreenLines + LineOffset[row],
-            ScreenLines + LineOffset[row] + col))
-        || (enc_utf8 && ScreenLines[LineOffset[row] + col] == 0)))
+      && (enc_utf8 && ScreenLines[LineOffset[row] + col] == 0)) {
```
This condition could be further simplified (`has_byte ` and `enc_utf8` are `true`).

oni-link commented:
`eval.txt` contains still some uses of `'encoding'` in `json_decode`, `json_encode` and `string`.

oni-link commented on runtime/doc/options.txt:
```diff
@@ -2133,43 +2126,8 @@ A jump table for the options with a short description can be found at |Q_op|.
-
-	The character encoding of files can be different from 'encoding'.
-	This is specified with 'fileencoding'.  The conversion is done with
-	iconv() or as specified with 'charconvert'.
```
Is this block not helpful anymore?

bfredl replied:
Looks like it makes sense to keep it.

oni-link commented:
`options.txt` contains still some uses of `'encoding'` in `charconvert`, `fileencoding`, `guifontwide`.

bfredl commented:
Thanks.

 I also updated parts of `mbyte.txt`. Though quite a lot of that file is quite out-of-date or irrelevant to nvim already. I mostly only updated direct references to 'encoding' being detected from environment or configurable.  No claim of correctness or relevance is made of text I have not changed.

kranki commented on src/nvim/globals.h:
```diff
@@ -778,44 +778,15 @@ EXTERN int vr_lines_changed INIT(= 0);      /* #Lines changed by "gR" so far */
-EXTERN int has_mbyte INIT(= 0);                 /* any multi-byte encoding */
+EXTERN int enc_dbcs INIT(= 0);       ///< Deprecated: always zero
+EXTERN bool enc_utf8 INIT(= true);   ///< UTF-8 encoded Unicode (always true)
+EXTERN bool has_mbyte INIT(= true);  ///< Any multi-byte encoding (always true)
```
Could these be `const`. Maybe some tool could then point dead code. At least flexelint (PC-lint) can but it isn't free.

bfredl replied:
let's make them `#define` and see what happens. Maybe we get too many dead code warnings, gcc doesn't, let see what clang says. Anyway "grep" might be enough as a tool to find dead code :)

bfredl commented:
@justinmk would you be fine merging this or would you want to take a look?

justinmk commented:
Should we release 0.1.6 first?

bfredl commented:
We could, but any particular reason for not having this in 0.1.6 ?

justinmk commented:
Just that we have enough goodies already for 0.1.6, and since end-users won't notice this change, why not let it bake on master for a bit?

justinmk commented:
(To be clear: we could release 0.1.6 tomorrow, then merge this)

bfredl commented:
Aha, you mean that you are going to release 0.1.6 real soon? That makes sense then.

bfredl commented:
But there is one change I think we should get into 0.1.6, make `nvim_x_set_var` and `del_var` not returning the old value as discussed in #4411, as releasing 0.1.6 is releasing the new `nvim_` api.

justinmk commented:
Ok, agreed.

justinmk commented on runtime/doc/mbyte.txt:
```diff
@@ -314,8 +307,8 @@ The {name} can be any encoding name that your system supports.  It is passed
 to iconv() to convert between the encoding of the file and the current locale.
```
to iconv() to convert between UTF-8 and the encoding of the file.

justinmk commented on runtime/doc/options.txt:
```diff
@@ -1077,8 +1074,7 @@ A jump table for the options with a short description can be found at |Q_op|.
 	This option lets you choose which characters might cause a line
-	break if 'linebreak' is on.  Only works for ASCII and also for 8-bit
-	characters when 'encoding' is an 8-bit encoding.
+	break if 'linebreak' is on.  Only works for ASCII characters.
```
utf8 is an 8-bit encoding so this should say

    Works for ASCII and 8-bit characters.

(or maybe just delete the "works for ASCII" sentence.)

bfredl replied:
No, utf-8 is a (drumroll please) multibyte encoding. If it was an 8bit encoding (by vim definition) every 'encoding' would be an 8bit encoding.

justinmk commented on runtime/doc/options.txt:
```diff
@@ -2133,44 +2124,13 @@ A jump table for the options with a short description can be found at |Q_op|.
-	shada file, etc.  It sets the kind of characters which Vim can work
-	with.  See |encoding-names| for the possible values.
+'encoding' 'enc'	Removed. |vim-differences| {Nvim}
+	Nvim always uses UTF-8 internally.
```
I think it's worth keeping the mention of RPC/remote as well:

    Nvim always uses UTF-8 internally. RPC communication 
    (remote plugins/GUIs) must use UTF-8 strings.

justinmk commented on test/functional/ex_cmds/encoding_spec.lua:
```diff
@@ -15,27 +15,16 @@ describe('&encoding', function()
     execute('set encoding=latin1')
     -- error message expected
     feed('<cr>')
-    neq(nil, string.find(eval('v:errmsg'), '^E905:'))
```
check for `E474` instead?

justinmk commented on test/functional/ex_cmds/encoding_spec.lua:
```diff
@@ -15,27 +15,16 @@ describe('&encoding', function()
-    execute('set encoding=utf-8')
-    -- error message expected
+  it('cannot be changed before startup', function()
+    clear('set enc=latin1')
```
check for `E474`

justinmk commented on test/functional/shada/history_spec.lua:
```diff
@@ -173,158 +171,57 @@ describe('ShaDa support code', function()
-    reset()
     funcs.setline('.', {'«\171'})
     nvim_command('&')
+    --nvim_command('substitute/«/./ge')
```
`:&` does not repeat flags so the comment should be like

    nvim_command('&')  -- :substitute/«/./

bfredl replied:
oops, this was probably leftover from debugging...

justinmk commented on test/functional/shada/history_spec.lua:
```diff
@@ -173,158 +171,57 @@ describe('ShaDa support code', function()
   function()
-    -- \171 is U+00AB LEFT-POINTING DOUBLE ANGLE QUOTATION MARK in latin1
-    nvim_command('silent! /«/')
+    nvim_command('silent! /\171/')
```
perhaps the `-- \171 is U+00AB LEFT-POINTING ...` comment should be kept in this case

justinmk commented on test/functional/shada/registers_spec.lua:
```diff
@@ -128,36 +128,24 @@ describe('ShaDa support code', function()
   end)
 
-  it('dumps and loads history correctly when &encoding /= UTF-8 when dumping',
+  it('dumps and loads history correctly with 8-bit contents',
```
Maybe "8-bit single-byte" is clearer than "8-bit" in these test descriptions.

justinmk commented on src/nvim/mbyte.c:
```diff
@@ -1,68 +1,27 @@
+/// read-only and always reads "utf-8".
+///
+/// The cell width on the display needs to be determined from the character
+/// value. Recognizing bytes is easy: 0xxx.xxxx is a single-byte char,
```
"Recognizing UTF-8 bytes..."

justinmk commented on src/nvim/mbyte.c:
```diff
@@ -1,68 +1,27 @@
+///
+/// utf-8 is used everywhere in the core. This is in registers, text
+/// manipulation, buffers, etc. Nvim core communicates with external plugins
+/// and gui:s in this encoding.
```
GUIs

justinmk commented on src/nvim/mbyte.h:
```diff
@@ -28,6 +28,18 @@
 #define ENC_LATIN9     0x400       /* Latin9 */
 #define ENC_MACROMAN   0x800       /* Mac Roman (not Macro Man! :-) */
 
+// TODO(bfredl): eventually we should keep only one of the namings
```
`mb_` would be the least churn, and helps with comparing vim-patches.

bfredl replied:
Agree, it is also matches the name of the module/feature ("multibyte"), and utf is an awkward abbreviation for utf8 :)

justinmk commented on src/nvim/fileio.c:
```diff
@@ -4165,9 +4165,8 @@ static bool need_conversion(const char_u *fenc)
-    /* Specified encoding matches with 'encoding'.  This requires
-     * conversion when 'encoding' is Unicode but not UTF-8. */
-    return enc_unicode != 0;
+    // Specified encoding matches with 'encoding'.
```
"...matches utf-8"

justinmk commented:
LGTM

justinmk commented on runtime/doc/mbyte.txt:
```diff
@@ -1245,10 +1226,10 @@ not everybody is able to type a composing character.
-'encoding'	Encoding used for the keyboard and display.  It is also the
-		default encoding for files.
+'encoding'	The encoding used internally used to be configurable with
+		this option. Now this is always "utf-8"
```
'encoding'	Internal string encoding, always "utf-8".

bfredl commented:
Thanks for comments, updated.

bfredl commented:
Rebased. Any more comments before I merge this?

jamessan commented on runtime/doc/mbyte.txt:
```diff
@@ -70,29 +70,24 @@ See |mbyte-locale| for details.
-If you are working in a terminal (emulator) you must make sure it accepts the
-same encoding as which Vim is working with.
+If you are working in a terminal (emulator) you must make sure it accepts
+UTF-8, the encoding which Vim is working with. Otherwise only ASCII can
```
This rewording seems odd.  Why is that necessary?  Shouldn't `'termencoding'` define the input/output encoding, thus allowing nvim to still work in non-utf8 environments?

It seems like the overall change here is essentially what I [proposed] a few years ago to Bram, except we're missing the part where the environment's locale sets the default value for `'termencoding'` rather than `'encoding'`.

[proposed]: https://www.mail-archive.com/vim_dev@googlegroups.com/msg05992.html

justinmk replied:
I thought we punted 'termencoding' long ago. As for wording, it could be tightened up:

    If you are working in a terminal (emulator) it must accept UTF-8. |'encoding'|

bfredl replied:
It is not odd, it just adds relevant information. `'termencoding'`was removed #2631 a while ago, and was non-functional even longer before (since the first big tui refactor, I think). Given the complete lack of complaints of the form "nvim is broken in my latin1 terminal!" since we stopped detecting `'encoding'` from the envirionment, I don't think adding it back will be neccessary.

justinmk commented on src/nvim/option.c:
```diff
@@ -2603,17 +2595,12 @@ did_set_string_option (
-        errmsg = mb_init();
-        redraw_titles();
+        // only encoding=utf-8 allowed
+        if (STRCMP(p_enc, "utf-8") != 0) {
```
Should this check `p`? Or was `p_enc` already set to an invalid value by this point?
